scripts/subspace_mlp_demo.py

```diff
@@ -0,0 +1,191 @@
+    velocity = hyperparams["beta_2"] * velocity + (1.0 - hyperparams["beta_2"]) * (grads ** 2.0)
+    # Bias correction
+    hat_mass = mass / (1 - hyperparams["beta_1"])
+    hat_velocity = velocity / (1 - hyperparams["beta_2"])
+    # Update
+    params = params - hyperparams["lr"] / (jnp.sqrt(hat_velocity) + hyperparams["epsilon"]) * hat_mass
+    return params, mass, velocity
+
+
+def generate_projection(key, d, D):
+    M = random.normal(key, shape=(d,D))
+    M = M / jnp.linalg.norm(M, axis=-1, keepdims=True)
+    return M
+
+
+def subspace_learning(key, model, datasets, d, hyperparams, n_epochs=300):
+    key_params, key_subspace = random.split(key)
+    _, num_features = train_ds["image"].shape
+
+    x0 = jnp.zeros(num_features)
+    w_init = model().init(key_params, x0)["params"]
+    w_init_flat, reconstruct_fn = jax.flatten_util.ravel_pytree(w_init)
+
+    D = len(w_init_flat)
+    A = generate_projection(key_subspace, d, D)
+    projected_loss_partial = partial(projected_loss, M=A,
+                                     reconstruct_fn=reconstruct_fn,
+                                     flat_params0=w_init_flat)
+    loss_grad_wrt_theta = jax.grad(projected_loss_partial)
+
+    theta = jnp.zeros((1, d))
+    mass = jnp.zeros((1, d))
+    velocity = jnp.zeros((1, d))
+
+    for e in range(n_epochs):
+        grads = loss_grad_wrt_theta(theta, datasets["train"])
+        theta, mass, velocity = adam_update(grads, theta, mass, velocity, hyperparams)
+
+        params_now = theta_to_flat_params(theta, A, w_init_flat)
+        params_now = reconstruct_fn(params_now)
+
+        epoch_loss = normal_loss(params_now, datasets["train"])
+        epoch_accuracy = normal_accuracy(params_now, datasets["train"])
+        if e % 100 == 0 or e == n_epochs - 1:
+            end = "\n"
+            epoch_val_accuracy = normal_accuracy(params_now, datasets["test"])
+            val_str = f" || val acc: {epoch_val_accuracy:0.2%}"
+        else:
+            end = "\r"
+            val_str = ""
+            
+        metric_str = f"epoch: {e+1:03} || acc: {epoch_accuracy:0.2%} || loss:{epoch_loss:0.2f}"
+        metric_str += val_str
+        print(metric_str, end=end)
+
+    return theta, epoch_val_accuracy
+
+
+if __name__ == "__main__":
+    plt.rcParams["axes.spines.right"] = False
+    plt.rcParams["axes.spines.top"] = False
+
+    key = random.PRNGKey(314)
+    n_features = 28 ** 2
+    train_ds, test_ds = get_datasets()
+    train_ds["image"] = train_ds["image"].reshape(-1, n_features)
+    test_ds["image"] = test_ds["image"].reshape(-1, n_features)
+
+    datasets = {
+        "train": train_ds,
+        "test": test_ds,
+    }
+
+    hyperparams = {
+        "lr": 1e-1,
+        "beta_1": 0.9,
+        "beta_2": 0.999,
+        "epsilon": 1e-7
+    }
+
+    min_dim, max_dim = 10, 1300
+    jump_size = 50
+    subspace_dims = [2] + list(range(min_dim, max_dim, jump_size))
+
+    acc_vals = []
+    n_epochs = 300
+    for dim in subspace_dims:
```

Comment:
can you vmap this?  (WIll probably need to jit `subspace_learning` first)

Reply:
AFAIK it is not possible to vmap over subspaces of different dimension size. Doing so implies varying the dimension of `params_subspace` incurring in an error of the form :

```python
TypeError: Shapes must be 1D sequences of concrete values of integer type
```

See "functions with argument-value dependent shapes" inside the [Common Gotchas](https://jax.readthedocs.io/en/latest/notebooks/Common_Gotchas_in_JAX.html#python-control-flow-jit).

As an example, we can consider the following code

```python
In [1]: import jax

In [2]: import jax.numpy as jnp

In [3]: def project_test(d, D):
   ...:     key = jax.random.PRNGKey(314)
   ...:     return jax.random.normal(key, shape=(D, d))
   ...: 

In [5]: from functools import partial

In [6]: project_partial = partial(project_test, D=10)

In [7]: project_partial(2)
Out[7]: 
DeviceArray([[ 0.06268313, -0.1967512 ],
             [-0.15028353,  0.07998917],
             [ 2.2091527 , -0.04232812],
             [-0.8189614 ,  1.0454589 ],
             [ 0.7797299 , -0.02449282],
             [-2.4424412 ,  0.9088825 ],
             [-1.2232977 ,  1.436878  ],
             [-0.4423918 , -0.53059614],
             [ 0.8967031 ,  1.8957794 ],
             [ 0.6805559 ,  0.32822272]], dtype=float32)

In [8]: d_values = jnp.array([2,3,4])

In [9]: jax.vmap(project_partial)(d_values)
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
<ipython-input-9-78f2e710b7a1> in <module>
----> 1 jax.vmap(project_partial)(d_values)

    [... skipping hidden 3 frame]

<ipython-input-3-4800702d72e8> in project_test(d, D)
      1 def project_test(d, D):
      2     key = jax.random.PRNGKey(314)
----> 3     return jax.random.normal(key, shape=(D, d))
      4 

~/miniconda3/lib/python3.9/site-packages/jax/_src/random.py in normal(key, shape, dtype)
    467                      f"got {dtype}")
    468   dtype = dtypes.canonicalize_dtype(dtype)
--> 469   shape = core.as_named_shape(shape)
    470   return _normal(key, shape, dtype)  # type: ignore
    471 

    [... skipping hidden 2 frame]

~/miniconda3/lib/python3.9/site-packages/jax/core.py in canonicalize_shape(shape, context)
   1441   except TypeError:
   1442     pass
-> 1443   raise _invalid_shape_error(shape, context)
   1444 
   1445 def canonicalize_dim(d: DimSize, context: str="") -> DimSize:

TypeError: Shapes must be 1D sequences of concrete values of integer type, got (10, Traced<ShapedArray(int32[])>with<BatchTrace(level=1/0)>
  with val = DeviceArray([2, 3, 4], dtype=int32)
       batch_dim = 0).
If using `jit`, try using `static_argnums` or applying `jit` to smaller subfunctions.
```

Reply:
You are right, we cannot easily use vmap. Let's stick to for loops ;)